### PR TITLE
split CallAPI

### PR DESCRIPTION
**Pull request by marcaudefroy**
Signed-off-by: Marc Audefroy <marc.audefroy@corp.ovh.com>

### CODE REVIEW

File: ovh/ovh.go [hunk 1]
```diff
@@ -251,39 +251,22 @@ var getEndpointForSignature = func(c *Client) string {
...
-//
-// Call will automatically assemble the target url from the endpoint
-// configured in the client instance and the path argument. If the reqBody
-// argument is not nil, it will also serialize it as json and inject
-// the required Content-Type header.
-//
-// If everyrthing went fine, unmarshall response into resType and return nil
-// otherwise, return the error
-func (c *Client) CallAPI(method, path string, reqBody, resType interface{}, needAuth bool) error {
+//CreateRequest construct the request for OVH api
```
Comment (gregdel) [inline ovh/ovh.go]: Space before the comment ? :)
construct -> constructs (or even returns)

File: ovh/ovh.go [hunk 2]
```diff
@@ -251,39 +251,22 @@ var getEndpointForSignature = func(c *Client) string {
...
+func (c *Client) CreateRequest(method, path string, reqBody interface{}, needAuth bool) (*http.Request, error) {
```
Comment (gregdel) [inline ovh/ovh.go]: CreateRequest -> NewRequest ?

File: ovh/ovh.go [hunk 3]
```diff
@@ -321,6 +304,13 @@ func (c *Client) CallAPI(method, path string, reqBody, resType interface{}, need
 
 	// Send the request with requested timeout
 	c.Client.Timeout = c.Timeout
+
+	return req, nil
+}
+
+//ExecuteRequest and unmarshall response into resType
```
Comment (gregdel) [inline ovh/ovh.go]: Space and 
ExecuteRequest => Do
Just like the http library ?

File: ovh/ovh.go [hunk 4]
```diff
@@ -321,6 +304,13 @@ func (c *Client) CallAPI(method, path string, reqBody, resType interface{}, need
...
+func (c *Client) ExecuteRequest(req *http.Request, resType interface{}) error {
+	c.Client.Timeout = c.Timeout
```
Comment (gregdel) [inline ovh/ovh.go]: Why would you update the timeout here, it should already be set

**Comment by marcaudefroy** [comment]
Updated and rebased.

File: ovh/ovh.go [hunk 5]
```diff
@@ -321,6 +304,12 @@ func (c *Client) CallAPI(method, path string, reqBody, resType interface{}, need
 
 	// Send the request with requested timeout
 	c.Client.Timeout = c.Timeout
+
+	return req, nil
+}
+
+// DoRequest and unmarshall response into resType
+func (c *Client) DoRequest(req *http.Request, resType interface{}) error {
```
Comment (gregdel) [inline ovh/ovh.go]: Why not just `Do` ?
This should almost be like a sentence: c.Do(req) == `client do request` and not `client do  request request`

**Comment by gregdel** [comment]
That's a big diff in the tests, are you sure we need that many lines of diff ? The commit message does not really explain your intentions.

**Comment by marcaudefroy** [comment]
Just refactoring of piece of code that test request values and reuse it on tests of NewRequest method.

**Comment by gregdel** [comment]
Refactoring should be made in a separate PR.

**Comment by marcaudefroy** [comment]
done

**Comment by PouuleT** [comment]
Now that we have a better split, why not do something even more similar to the http lib : 
NewRequest -> Do -> Unmarshal the response
Maybe something like that ?

```
func (c *Client) CallAPI(method, path string, reqBody, resType interface{}, needAuth bool) error {
￼ 	req, err := c.NewRequest(method, path, reqBody, needAuth)
￼ 	if err != nil {
￼ 		return err
￼ 	}
￼ 	response, err := c.Do(req)
 ￼ 	if err != nil {
￼ 		return err
￼ 	}
        return c.UnmarshalResponse(resType)
￼ }
```

**Comment by marcaudefroy** [comment]
Hum.. why not but Do() becomes just. Ok for this?
```
func (c *Client) Do(req *http.Request, resType interface{}) error {
	return c.Client.Do(req)
}
```

**Comment by marcaudefroy** [comment]
Thanks :)

**Pull request merged**